larpandoracontent/LArCheating/CheatingParticleIdBaseTool.cc

```diff
@@ -0,0 +1,107 @@
+
+namespace lar_content
+{
+
+CheatingParticleIdBaseTool::CheatingParticleIdBaseTool() 
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void CheatingParticleIdBaseTool::GetTargetParticleWeight(const PfoList *const pPfoList, const bool objectOwnedByMaster, float &targetParticleWeight, float &totalWeight, std::function<bool(const MCParticle *const)> fCriteria)
+{
+    targetParticleWeight = 0.f; totalWeight = 0.f;
+
+    for (const ParticleFlowObject *const pPfo : *pPfoList)
+    {
+        ClusterList twoDClusters;
+        LArPfoHelper::GetTwoDClusterList(pPfo, twoDClusters);
+
+        CaloHitList caloHitList;
+
+        for (const Cluster *const pCluster : twoDClusters)
+        {
+            const HitType hitType(LArClusterHelper::GetClusterHitType(pCluster));
+
+            if ((TPC_VIEW_U != hitType) && (TPC_VIEW_V != hitType) && (TPC_VIEW_W != hitType))
+                throw StatusCodeException(STATUS_CODE_INVALID_PARAMETER);
+
+            pCluster->GetOrderedCaloHitList().FillCaloHitList(caloHitList);
+        }
+
+        for (const CaloHit *const pCaloHit : caloHitList)
+        {
+            float thisTargetParticleWeight = 0.f, thisTotalWeight = 0.f;
+            CheatingParticleIdBaseTool::GetTargetParticleWeight(pCaloHit, objectOwnedByMaster, thisTargetParticleWeight, thisTotalWeight, fCriteria);
+
+            targetParticleWeight += thisTargetParticleWeight;
+            totalWeight += thisTotalWeight;
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void CheatingParticleIdBaseTool::GetTargetParticleWeight(const CaloHit *const pCaloHit, const bool objectOwnedByMaster, float &targetParticleWeight, float &totalWeight, std::function<bool(const MCParticle *const)> fCriteria)
+{
+    targetParticleWeight = 0.f; totalWeight = 0.f;
+
+    const CaloHit *const pCaloHitMaster(objectOwnedByMaster ? pCaloHit : static_cast<const CaloHit *>(pCaloHit->GetParentAddress()));
+    const MCParticleWeightMap &hitMCParticleWeightMap(pCaloHitMaster->GetMCParticleWeightMap());
+
+    if (hitMCParticleWeightMap.empty())
+        return;
+
+    MCParticleList mcParticleList;
+    for (const auto &mapEntry : hitMCParticleWeightMap) mcParticleList.push_back(mapEntry.first);
+    mcParticleList.sort(LArMCParticleHelper::SortByMomentum);
+
+    for (const MCParticle *const pMCParticle : mcParticleList)
+    {
+        const float weight(hitMCParticleWeightMap.at(pMCParticle));
+        const MCParticle *const pParentMCParticle(LArMCParticleHelper::GetParentMCParticle(pMCParticle));
+
+        if (fCriteria(pParentMCParticle))
+            targetParticleWeight += weight;
+
+        totalWeight += weight;
+    }
+
+    // ATTN normalise arbitrary input weights at this point
+    if (totalWeight > std::numeric_limits<float>::epsilon())
```

Comment:
Did you need to change any of these functions? There's only this last assignment of the weights that look different.

Reply:
The only difference in these functions is that the neutrino labels have been changed to target particle labels and the IsNeutrino function replaced by fCriteria.